README.md
# SwiftyTextTable

A lightweight Swift library for generating text tables.

[![Build Status](https://travis-ci.org/scottrhoyt/SwiftyTextTable.svg?branch=master)](https://travis-ci.org/scottrhoyt/SwiftyTextTable)
[![codecov.io](https://codecov.io/github/scottrhoyt/SwiftyTextTable/coverage.svg?branch=master)](https://codecov.io/github/scottrhoyt/SwiftyTextTable?branch=master)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![Swift Package Manager compatible](https://img.shields.io/badge/Swift%20Package%20Manager-compatible-brightgreen.svg)](https://github.com/apple/swift-package-manager)
[![CocoaPods](https://img.shields.io/cocoapods/v/SwiftyTextTable.svg)]()
![Platform OS X + Linux](https://img.shields.io/badge/Platform-OS%20X%20%2B%20Linux-blue.svg)
[![Language Swift 4.0](https://img.shields.io/badge/Language-Swift%204.0-orange.svg)](https://swift.org)

![Example](http://i.imgur.com/utoa6TK.png)

## Swift Language Support

SwiftyTextTable is now Swift 4.0 compatible! The last release to support Swift
3.1 was [0.7.1](https://github.com/scottrhoyt/SwiftyTextTable/releases/tag/0.7.1).
The last release to support Swift 2.3 was [0.3.1](https://github.com/scottrhoyt/SwiftyTextTable/releases/tag/0.3.1).

## Installation

### Carthage (OS X)
You can use [Carthage](https://github.com/Carthage/Carthage) to install
SwiftyTextTable by adding it to your `Cartfile`:

```
github "scottrhoyt/SwiftyTextTable"
```

### Swift Package Manager (OS X + Linux)
You can use [The Swift Package Manager](https://swift.org/package-manager) to
install SwiftyTextTable by adding the proper description to your
`Package.swift` file:

```swift
import PackageDescription

let package = Package(
    name: "<YOUR_PROJECT_NAME>",
    dependencies: [
        .package(url: "https://github.com/scottrhoyt/SwiftyTextTable.git", from: "0.5.0")
    ]
)
```

### CocoaPods (OS X)

You can use [CocoaPods](https://cocoapods.org) to install SwiftyTextTable by
adding it to your `Podfile`:

```ruby
pod 'SwiftyTextTable'
```

### Manual

Simply copy the `*.swift` files from the `Source/SwiftyTextTable` directory into
your project.

## Usage

```swift
import SwiftyTextTable

// First create some columns
let foo = TextTableColumn(header: "foo")
let bar = TextTableColumn(header: "bar")
let baz = TextTableColumn(header: "baz")

// Then create a table with the columns
var table = TextTable(columns: [foo, bar, baz])

// Then add some rows
table.addRow([1, 2, 3])
table.addRow([11, 22, 33])

// Then render the table and use
let tableString = table.render()
print(tableString)

/*
+-----+-----+-----+
| foo | bar | baz |
+-----+-----+-----+
| 1   | 2   | 3   |
| 11  | 22  | 33  |
+-----+-----+-----+
*/

// Put a header on the table if you'd like
table.header = "my foo table"
print(table.render())

/*
+-----------------+
| my foo table    |
+-----------------+
| foo | bar | baz |
+-----+-----+-----+
| 1   | 2   | 3   |
| 11  | 22  | 33  |
+-----+-----+-----+
*/
```

Any `CustomStringConvertible` can be used for row `values`.

### Creating Tables from Arrays of Objects with `TextTableRepresentable`

Let's say you have an array of objects that looks this:

```swift
enum AnimalType: String, CustomStringConvertible {
    case dog = "Dog"
    case cat = "Cat"
    case gorilla = "Gorilla"

    var description: String {
        return self.rawValue
    }
}

struct Pet {
    let type: AnimalType
    let name: String
    let canHazPizza: Bool
}

let furball = Pet(type: .cat, name: "Furball", canHazPizza: false)
let bestFriend = Pet(type: .dog, name: "Best Friend", canHazPizza: true)
let scary = Pet(type: .gorilla, name: "Scary", canHazPizza: true)
let pets = [furball, bestFriend, scary]
```

Now you want to print a table containing your `pets`. You can accomplish this
by having `Pet` conform to `TextTableRepresentable`:

```swift
extension Pet: TextTableRepresentable {
    static var columnHeaders: [String] {
        return ["Name", "Animal", "Can Haz Pizza?"]
    }

    var tableValues: [CustomStringConvertible] {
        return [name, type, canHazPizza ? "yes" : "no"]
    }

    // Optional
    static var tableHeader: String? {
      return "My Pets"
    }
}
```

You can now print a table of your `pets` simply:

```swift
print(pets.renderTextTable())

/*
+----------------------------------------+
| My Pets                                |
+----------------------------------------+
| Name        | Animal  | Can Haz Pizza? |
+-------------+---------+----------------+
| Furball     | Cat     | no             |
| Best Friend | Dog     | yes            |
| Scary       | Gorilla | yes            |
+-------------+---------+----------------+
*/
```

### Fence Custimization

You can also customize the output of `TextTable.render()` by using different
values for `columnFence`, `rowFence`, and `cornerFence`.

```swift
table.columnFence = ":"
table.rowFence = "."
table.cornerFence = "."

print(table.render())

/*
...................
: foo : bar : baz :
...................
: 1   : 2   :     :
: 11  : 22  : 33  :
...................
*/
```

### Row Padding/Truncation

When adding rows, `TextTable` will automatically pad the rows with empty strings
when there are fewer `values` than columns. `TextTable` will also disregard all
`values` over the column count.

```swift
let foo = TextTableColumn(header: "foo")
let bar = TextTableColumn(header: "bar")
let baz = TextTableColumn(header: "baz")

var table = TextTable(columns: [foo, bar, baz])

table.addRow([1, 2])
table.addRow([11, 22, 33])
table.addRow([111, 222, 333, 444])

let tableString = table.render()
print(tableString)

/*
+-----+-----+-----+
| foo | bar | baz |
+-----+-----+-----+
| 1   | 2   |     |
| 11  | 22  | 33  |
| 111 | 222 | 333 |
+-----+-----+-----+
*/
```

### Console Formatting Support

SwiftyTextTable will recognize many console escape sequences used to format
output (e.g. [Rainbow](https://github.com/onevcat/Rainbow)) and account for them
in constructing the table.

### API Reference

Check out the full API reference [here](http://scottrhoyt.github.io/SwiftyTextTable/).

## License

SwiftyTextTable is released under the [MIT License](https://github.com/scottrhoyt/SwiftyTextTable/blob/master/LICENSE).


CHANGELOG.md
## 0.9.0

##### Breaking

* None

##### Enhancements

* Pre-compute column width to increase rendering performance.

##### Bug Fixes

* Fix column width when using terminal escape sequences such as coloring.

## 0.8.2

##### Breaking

* None

##### Enhancements

* None.

##### Bug Fixes

* Fix Swift Package Manager manifest to include SwiftyTextTable library product.

## 0.8.1

##### Breaking

* None

##### Enhancements

* Update package manifest to v4 format.

##### Bug Fixes

* None

## 0.8.0

##### Breaking

* Update to Swift 4.

##### Enhancements

* None

##### Bug Fixes

* None

## 0.7.1

##### Breaking

* None

##### Enhancements

* None

##### Bug Fixes

* Fix Swift 4.0.2 deprecation of String.characters

## 0.7.0

##### Breaking

* None

##### Enhancements

* CI tests for Xcode 9 and Swift 4 on Linux support.

##### Bug Fixes

* Fixes for Xcode 9 beta 3

## 0.6.0: Table Service

##### Breaking

* `TextTableColumn.width` was refactored to `TextTableColumn.width()` to better
  reflect the O(n) complexity of calculating this value.
* `TextTableObject` was renamed to `TextTableRepresentable`.

##### Enhancements

* Added Documentation.
* Re-enabled console formatting support for Linux.
* Improved compilation time.
* Added convenience method for rendering tables from an array of
  `TextTableRepresentable`s (`renderTextTable`).
* Now available via CocoaPods.

##### Bug Fixes

* None

## 0.5.0: Heading In The Right Direction

##### Breaking

* `TableObject.tableHeaders` was rename `TableObject.columnHeaders` for clarity.

##### Enhancements

* `TextTable`s can now optionally have table headers.
* `TableObject`s can now optionally supply a table header

##### Bug Fixes

* None

## 0.4.0: Swift 3.0 Support

##### Breaking

* `SwiftyTextTable` adopts the new Swift 3.0 convention of explicit first
  parameter labels for functions.
* Linux support for dealing with console formatting escape sequences has been
  removed for the time being due to regular expression portability problems.

##### Enhancements

* Swift 3.0 support!

##### Bug Fixes

* None

## 0.3.1: Swift 2.3 Support

##### Breaking

* None

##### Enhancements

* SwiftyTextTable is now compatible with Swift 2.3 on OS X and Linux thanks to
  the great work of [Norio Nomura](https://github.com/norio-nomura).

##### Bug Fixes

* None

## 0.3.0: Table Objects Oh My!

##### Breaking

* `TextTable.addRow` now takes an array instead of an argument list. This was to
  support `TextTableObject`.

##### Enhancements

* `TextTableObject` allows for easy table creation from an array of objects.
* Better SPM testing support.

##### Bug Fixes

* None

## 0.2.2: No Escape

##### Breaking

* None.

##### Enhancements

* None.

##### Bug Fixes

* Fix length calculation for string with console escape sequences.  
  [Scott Hoyt](https://github.com/scottrhoyt)
  [#5](https://github.com/scottrhoyt/SwiftyTextTable/issues/5)


Source/SwiftyTextTableTests/SwiftyTextTableTests.swift
//
//  SwiftyTextTableTests.swift
//  SwiftyTextTableTests
//
//  Created by Scott Hoyt on 2/3/16.
//  Copyright © 2016 Scott Hoyt. All rights reserved.
//

import XCTest
import SwiftyTextTable

class SwiftyTextTableTests: XCTestCase {

    func fooTable() -> TextTable {
        let foo = TextTableColumn(header: "foo")
        let bar = TextTableColumn(header: "bar")
        let baz = TextTableColumn(header: "baz")
        var table = TextTable(columns: [foo, bar, baz])
        table.addRow(["1", "2"])
        table.addRow([11, 22, 33])
        table.addRow([111, 222, 333, 444])
        return table
    }

    func testRenderDefault() {
        let output = fooTable().render()
        let expected = "+-----+-----+-----+\n" +
                       "| foo | bar | baz |\n" +
                       "+-----+-----+-----+\n" +
                       "| 1   | 2   |     |\n" +
                       "| 11  | 22  | 33  |\n" +
                       "| 111 | 222 | 333 |\n" +
                       "+-----+-----+-----+"
        XCTAssertEqual(output, expected)
    }

    func testRenderCustom() {
        var table = fooTable()
        table.columnFence = "!"
        table.rowFence = "*"
        table.cornerFence = "."
        let output = table.render()
        let expected = ".*****.*****.*****.\n" +
                       "! foo ! bar ! baz !\n" +
                       ".*****.*****.*****.\n" +
                       "! 1   ! 2   !     !\n" +
                       "! 11  ! 22  ! 33  !\n" +
                       "! 111 ! 222 ! 333 !\n" +
                       ".*****.*****.*****."
        XCTAssertEqual(output, expected)
    }

    func testStripping() {
        let c1 = TextTableColumn(header: "\u{001B}[0mHello\u{001B}[0m")
        XCTAssertEqual(c1.width, 5)

        let c2 = TextTableColumn(header: "\u{001B}[31m\u{001B}[4;31;93mHello World\u{001B}[0m\u{001B}[0m")
        XCTAssertEqual(c2.width, 11)

        let c3 = TextTableColumn(header: "\u{001B}[0m\u{001B}[0m")
        XCTAssertEqual(c3.width, 0)

        let c4 = TextTableColumn(header: "\u{001B}[31mHello World\u{001B}[0m")
        XCTAssertEqual(c4.width, 11)

        let c5 = TextTableColumn(header: "\u{001B}[4;31;42;93;5mHello World\u{001B}[0m")
        XCTAssertEqual(c5.width, 11)

        let c6 = TextTableColumn(header: "\u{001B}[4;31;93mHello World\u{001B}[0m")
        XCTAssertEqual(c6.width, 11)

        let c7 = TextTableColumn(header: "Hello World")
        XCTAssertEqual(c7.width, 11)
    }

    func testTableObjects() {
        // swiftlint:disable:next nesting
        struct TableObject: TextTableObject {
            static var tableHeaders: [String] {
                return [ "foo", "bar", "baz"]
            }

            let foo: Int
            let bar: String
            let baz: Double

            var tableValues: [CustomStringConvertible] {
                return [foo, bar, baz]
            }
        }

        let objects = [
            TableObject(foo: 1, bar: "2", baz: 3),
            TableObject(foo: 11, bar: "22", baz: 33),
            TableObject(foo: 111, bar: "222", baz: 333)
        ]

        let output = TextTable(objects: objects).render()
        let expected = "+-----+-----+-------+\n" +
                       "| foo | bar | baz   |\n" +
                       "+-----+-----+-------+\n" +
                       "| 1   | 2   | 3.0   |\n" +
                       "| 11  | 22  | 33.0  |\n" +
                       "| 111 | 222 | 333.0 |\n" +
                       "+-----+-----+-------+"
        print(output)
        print(expected)
        XCTAssertEqual(output, expected)

        let emptyOutput = TextTable(objects: [TableObject]()).render()
        let emptyExpected = "++\n" +
                            "||\n" +
                            "++\n" +
                              "\n" +
                            "++"
        XCTAssertEqual(emptyOutput, emptyExpected)
    }

    // MARK: - protocol XCTestCaseProvider for SPM
    lazy var allTests: [(String, () throws -> Void)] = [
        ("testRenderDefault", self.testRenderDefault),
        ("testRenderCustom", self.testRenderCustom),
        ("testStripping", self.testRenderCustom)
    ]
}
